### PR TITLE
Async functions to await a full or empty queue

### PR DESCRIPTION
**Pull request by overhacked**
Allow awaiting a full or empty queue.

- `Queue.empty()` is useful for chaining futures
  in a multi-threaded program when some subsequent
  action, like a verification step, needs to be
  taken once the queue is emptied.
- `Queue.full()` can help manage backpressure so that
  an autoscaling system can monitor when the queue fills
  before it starts increasing the number of workers.

### CODE REVIEW

File: src/limited.rs [hunk 1]
```diff
@@ -105,6 +105,16 @@ impl<T> Queue<T> {
     pub fn available(&self) -> isize {
         self.available.get()
     }
+    /// Await until the queue is full.
+    pub async fn full(&self) {
+        let capacity = self.capacity().try_into().unwrap();
+        let _ = self.pop_semaphore.acquire_many(capacity).await.unwrap();
```
Comment (bikeshedder) [inline src/limited.rs]: Tokio semaphores are fair in nature. This `acquire_many` call will actually block all other `acquire` calls until the queue is full. If you were to call this on a queue and you never put in `capacity` objects this call never returns and other calls to `pop` will be blocked indefinitely.

File: src/limited.rs [hunk 2]
```diff
@@ -105,6 +105,16 @@ impl<T> Queue<T> {
...
+    }
+    /// Await until the queue is empty.
+    pub async fn empty(&self) {
+        let capacity = self.capacity().try_into().unwrap();
+        let _ = self.push_semaphore.acquire_many(capacity).await.unwrap();
```
Comment (bikeshedder) [inline src/limited.rs]: This `acquire_many` call has the same problem as in the `full()` method.

File: src/resizable.rs [hunk 3]
```diff
@@ -87,6 +94,38 @@ impl<T> Queue<T> {
...
+        if self.len() == self.capacity() {
+            return;
+        }
+        self.notify_full.subscribe().changed().await.unwrap();
+    }
+    /// Await until the queue is empty.
+    pub async fn empty(&self) {
+        loop {
+            let capacity = self.capacity().try_into().unwrap();
+            let _permit = self.push_semaphore.acquire_many(capacity).await.unwrap();
```
Comment (bikeshedder) [inline src/resizable.rs]: Just like in the `limited` implementation I don't think this is correct. I guess the only (and correct) way around it is to use a channel for both `full` and `empty` notifications.

**Review by bikeshedder — CHANGES_REQUESTED** [review]
This is a nice addition to the crate.

I just think the "simple" implementations of those methods using just `Semaphore::acquire_many` are wrong. I'd like to see a test case that actually checks if calling the `full` and `empty` method don't block concurrent `push` and `pop` calls. The tests would look like that:
- Full
   - Create a queue with capacity of 2
   - call `full`  (`tokio::spawn`)
   - call `push`
   - call `pop` and make sure it doesn't block
   - call `push` twice and check that the `full` call returned
- Empty
    - Create a queue with capacity of 2
    - call `push` twice
    - call `empty` (`tokio::spawn`)
    - call `push`
    - call `pop` and make sure it doesn't block
    - call `pop` twice and check that `empty` call returned

One more thing: I don't like the names `empty` and `full`. `empty` almost sounds like it is supposed to `clear` the queue. How about creating naming them `wait_empty`/`wait_full` or `on_empty`/`on_full` instead? That would make it very clear that the queues aren't modified but you're waiting for a specific event to happen.

**Comment by bikeshedder** [comment]
If you were to change the `deadqueue::atomic::Available` implementation so it returns the new (or old) value after `add` and `sub` you could use that to find out if the queue is full and/or empty. That way you could avoid calling `available_permits` and remove one possible race condition where the queue is empty for the fraction of a second but `Semaphore::available_permits` doesn't show that because of concurrent `push`/`pop` calls.

Even then you might miss an `empty`/`full` event, but it's less likely to happen than calling into `Semaphore::available_permits` outside of any atomic guarantees.

**Comment by overhacked** [comment]
Hi, read all these comments. Thanks for the feedback! Trying to find time to rework the PR accordingly.

**Comment by overhacked** [comment]
Finally circled back to this and updated the `wait_full()` and `wait_empty()` to use atomics instead of `Semaphore::available_permits()`. Thanks for your patience and not closing the PR.

**Comment by bikeshedder** [comment]
Sorry for taking so long. I completely forgot about that PR. :see_no_evil: 

After reviewing it I only found two small details:

1. Every time you use `Available:add` you add 1 to the previous value to see if the capacity has been reached. I'd rather change that code so the `Available:add` function already returns the new value so this comparison doesn't need to happen. The same can be done for `Available::sub` where comparing the new value against `0` is much clearer than comparing the previous value against `1`.

2. It would be nice if there was a way to get a hold of the watch channel of `full` and `empty`. That way an application that wants to perform some task every time the queue is full or empty doesn't need to continuously recreate the channel receiver.

If you happen to find time to look into that I'll wait a bit. I'm just as happy to merge this PR without those changes.

File: src/resizable.rs [hunk 4]
```diff
@@ -133,6 +181,15 @@ impl<T> Queue<T> {
                     };
                     self.capacity.fetch_sub(1, Ordering::Relaxed);
                 }
+                // Currently this is a no-op because this resize() function
```
Comment (bikeshedder) [inline src/resizable.rs]: What do you mean by that? `resize()` takes a `&self` not a `&mut self`. :thinking:

**Comment by bikeshedder** [comment]
Oh, and please explain this comment to me...
- https://github.com/bikeshedder/deadqueue/pull/4/files#r1387924883

How is it borrowed mutably? :thinking:

**Comment by overhacked** [comment]
I should be able to look at these this weekend. Thanks!

**Comment by bikeshedder** [comment]
@overhacked Any updates on this PR?

**Comment by overhacked** [comment]
Requested changes pushed. Sorry I lost track of wrapping this up in the last few months of the year. Nothing going on in November or December... nothing else at all. :wink:

Thanks for your patience and support with this!

**Review by bikeshedder — APPROVED** [review]

**Comment by bikeshedder** [comment]
Thanks a lot. Awesome work! :tada:

**Comment by bikeshedder** [comment]
I just saw that I never actually released a new version including your changes. :facepalm: 

I just releated `deadqueue 0.2.5` on crates.io:

:package: https://crates.io/crates/deadqueue/0.2.5

Thanks again for this PR! :heart:

**Comment by overhacked** [comment]
Amazing, thank you!

**Pull request merged**